moshi/src/main/java/com/squareup/moshi/JsonWriter.kt

```diff
@@ -0,0 +1,548 @@
+   * or an array.
+   *  * Numbers may be [NaNs][Double.isNaN] or [infinities][Double.isInfinite].
+   *
+   *  Returns true if this writer has relaxed syntax rules.
+   */
+  @get:CheckReturnValue
+  public var isLenient: Boolean = false
+
+  /**
+   * Sets whether object members are serialized when their value is null. This has no impact on
+   * array elements. The default is false.
+   *
+   * Returns true if object members are serialized when their value is null. This has no impact on
+   * array elements. The default is false.
+   */
+  @get:CheckReturnValue
+  public var serializeNulls: Boolean = false
+
+  @JvmField
+  public var promoteValueToName: Boolean = false
+
+  /**
+   * Controls the deepest stack size that has begin/end pairs flattened:
+   *
+   *  - If -1, no begin/end pairs are being suppressed.
+   *  - If positive, this is the deepest stack size whose begin/end pairs are eligible to be flattened.
+   *  - If negative, it is the bitwise inverse (~) of the deepest stack size whose begin/end pairs have been flattened.
+   *
+   *
+   * We differentiate between what layer would be flattened (positive) from what layer is being
+   * flattened (negative) so that we don't double-flatten.
+   *
+   *
+   * To accommodate nested flattening we require callers to track the previous state when they
+   * provide a new state. The previous state is returned from [beginFlatten] and restored
+   * with [endFlatten].
+   */
+  @JvmField
+  public var flattenStackSize: Int = -1
+
+  private var tags: MutableMap<Class<*>, Any>? = null
+
+  /** Returns the scope on the top of the stack.  */
+  public fun peekScope(): Int {
+    check(stackSize != 0) { "JsonWriter is closed." }
+    return scopes[stackSize - 1]
+  }
+
+  /** Before pushing a value on the stack this confirms that the stack has capacity.  */
+  public fun checkStack(): Boolean {
+    if (stackSize != scopes.size) return false
+    if (stackSize == 256) {
+      throw JsonDataException("Nesting too deep at $path: circular reference?")
+    }
+    scopes = scopes.copyOf(scopes.size * 2)
+    pathNames = pathNames.copyOf(pathNames.size * 2)
+    pathIndices = pathIndices.copyOf(pathIndices.size * 2)
+    if (this is JsonValueWriter) {
+      stack = stack.copyOf(stack.size * 2)
+    }
+    return true
+  }
+
+  public fun pushScope(newTop: Int) {
+    scopes[stackSize++] = newTop
+  }
+
+  /** Replace the value on the top of the stack with the given value.  */
+  public fun replaceTop(topOfStack: Int) {
+    scopes[stackSize - 1] = topOfStack
+  }
+
+  /**
+   * Sets the indentation string to be repeated for each level of indentation in the encoded
+   * document. If `indent.isEmpty()` the encoded document will be compact. Otherwise the
+   * encoded document will be more human-readable.
+   *
+   * @param indent a string containing only whitespace.
+   */
+  public open fun setIndent(indent: String) {
+    this.indent = indent.ifEmpty { null }
```

Comment:
this can be inlined to a setter on the indent property right?

Reply:
Seems not. the param is nonnull but the field is nullable. If we need to make the field nonnull?

Reply:
I think you should try. It also should not be `open`

Reply:
As it was overrided in `JsonUtf8Writer`.

```java
  @Override
  public void setIndent(String indent) {
    super.setIndent(indent);
    this.separator = !indent.isEmpty() ? ": " : ":";
  }
```

Reply:
ah ok, fine to leave that for now

Reply:
Now I make it like this below, right?
```kotlin
public open var indent: String = ""
```
And the judgement in `JsonUtf8Writer`
```java
private void newline() throws IOException {
  if (getIndent().isEmpty()) {
    return;
  }
  ...
}
```

Reply:
I don't understand the second part of that questoin

Reply:
Here is the check of `indent` now: 


```java
private void newline() throws IOException {
  if (indent == null) {
    return;
  }

  sink.writeByte('\n');
  for (int i = 1, size = stackSize; i < size; i++) {
    sink.writeUtf8(indent);
  }
}
```

Reply:
Sorry I am still not sure what point you are trying to make here? Just leave the logic as it was or convert it to a protected open var property with a custom getter/setter. Preferably the latter